substrate/client/consensus/aura/src/import_queue.rs

```diff
@@ -346,14 +347,111 @@ where
+	})
+	.map_err(|e| sp_consensus::Error::Other(e.into()))?;
+	Ok(BasicQueue::new(
+		verifier,
+		Box::new(block_import.clone()),
+		justification_import,
+		spawner,
+		registry,
+	))
+}
+
+/// AURA block import.
+pub struct AuraBlockImport<Client, P: Pair + Clone, Block: BlockT, BI: BlockImport<Block> + Clone> {
+	block_import: BI,
+	authorities_tracker: Arc<AuthoritiesTracker<P, Block, Client>>,
+}
+
+impl<Client, P: Pair + Clone, Block: BlockT, BI: BlockImport<Block> + Clone>
+	AuraBlockImport<Client, P, Block, BI>
+where
+	Client: HeaderBackend<Block>
+		+ HeaderMetadata<Block, Error = sp_blockchain::Error>
+		+ ProvideRuntimeApi<Block>,
+	P::Public: Codec + Debug,
+	Client::Api: AuraApi<Block, AuthorityId<P>>,
+{
+	/// Create a new AURA block import.
+	pub fn new(
+		block_import: BI,
+		client: Arc<Client>,
+		compatibility_mode: &CompatibilityMode<NumberFor<Block>>,
+	) -> Result<(Self, Arc<AuthoritiesTracker<P, Block, Client>>), String> {
+		let authorities_tracker = Arc::new(AuthoritiesTracker::new(client, compatibility_mode)?);
+		Ok((
+			Self { block_import, authorities_tracker: authorities_tracker.clone() },
+			authorities_tracker,
+		))
+	}
+
+	/// Create a new AURA block import with an empty authorities tracker.
+	/// Usually you should _not_ use this method, as it will not have any initial authorities
+	/// imported. Use [`AuraBlockImport::new`] instead.
+	pub fn new_empty(
+		block_import: BI,
+		client: Arc<Client>,
+	) -> (Self, Arc<AuthoritiesTracker<P, Block, Client>>) {
+		let authorities_tracker = Arc::new(AuthoritiesTracker::new_empty(client));
+		(
+			Self { block_import, authorities_tracker: authorities_tracker.clone() },
+			authorities_tracker,
+		)
+	}
+}
+
+impl<Client, P: Pair + Clone, Block: BlockT, BI: BlockImport<Block> + Clone> Clone
+	for AuraBlockImport<Client, P, Block, BI>
+{
+	fn clone(&self) -> Self {
+		Self {
+			block_import: self.block_import.clone(),
+			authorities_tracker: self.authorities_tracker.clone(),
+		}
+	}
+}
+
+#[async_trait::async_trait]
+impl<
+		Client: Sync + Send,
+		P: Pair + Clone,
+		Block: BlockT,
+		BI: BlockImport<Block> + Send + Sync + Clone,
+	> BlockImport<Block> for AuraBlockImport<Client, P, Block, BI>
+where
+	Client: HeaderBackend<Block>
+		+ HeaderMetadata<Block, Error = sp_blockchain::Error>
+		+ ProvideRuntimeApi<Block>,
+	P::Public: Codec + Debug,
+	P::Signature: Codec,
+	Client::Api: AuraApi<Block, AuthorityId<P>>,
+	BI::Error: From<String>,
+{
+	type Error = BI::Error;
+
+	async fn check_block(
+		&self,
+		block: BlockCheckParams<Block>,
+	) -> Result<ImportResult, Self::Error> {
+		self.block_import.check_block(block).await
+	}
 
-	Ok(BasicQueue::new(verifier, Box::new(block_import), justification_import, spawner, registry))
+	/// Import a block.
+	async fn import_block(
+		&self,
+		block: BlockImportParams<Block>,
+	) -> Result<ImportResult, Self::Error> {
+		let post_header = block.post_header();
+		let res = self.block_import.import_block(block).await?;
```

Comment:
If the block was not successfully imported, we shouldn't import it into the authority tracker IMO.

Reply:
I think you're correct, otherwise we might get the authority tracker returning `Duplicate` errors. I think we should also not import on `ImportResult::AlreadyInChain` since even during state imports we will import authorities, so no need to re-import them. https://github.com/paritytech/polkadot-sdk/pull/9707/commits/ea8179de15e74f77ffc6eba8ed54e3968bdcd46c